pennylane/hf/tapering.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for tapering qubits using symmetries.
+"""
+from copy import deepcopy
+import numpy as np
+import pennylane as qml
+
+
+def _binary_matrix(terms, num_qubits):
```

Comment:
What is a binary matrix representation?

Reply:
We can represent Hamiltonian `H` as the summation of Pauli terms as `H   = \sum_{i}pP_{i}`. To find a binary matrix representation of H, we take each Pauli term `P_i` and represent it as a binary vector by storing the indices of PauliZ and PauliX operators in a vector of length 2*#qubits. Basically, first half has PauliZ and last half PauliX. Let's take the examples of  Z(0), X(1), and  Y(2) for 3 qubits:
```
Z(0) = [1, 0, 0, 0, 0, 0]
X(1) = [0, 0, 0, 0, 1, 0]
Y(2) = [0, 0, 1, 0, 0, 1]
```
Each of these vectors can be stacked to build a binary matrix representation.

Reply:
We need to have efficient functionalities for Hamiltonian arithmetics. Functions like this might be needed by the users for that purpose but not sure if users need this function now.

Reply:
> Functions like this might be needed by the users for that purpose but not sure if users need this function now.

It could be worth moving this to `pennylane/grouping`, and making it a public function. This will ensure it is easily found and available for when we start to use the Pauli grouping functions for Hamiltonian arithmetic